src/main/java/org/jenkinsci/plugins/workflow/cps/DSL.java

```diff
@@ -217,23 +222,37 @@ protected Object invokeStep(StepDescriptor d, Object args) {
      * @param args The arguments passed to the step.
      */
     protected Object invokeStep(StepDescriptor d, String name, Object args) {
-        final NamedArgsAndClosure ps = parseArgs(args, d);
-
-        CpsThread thread = CpsThread.current();
-
-        FlowNode an;
 
         // TODO: generalize the notion of Step taking over the FlowNode creation.
         boolean hack = d instanceof ParallelStep.DescriptorImpl || d instanceof LoadStep.DescriptorImpl;
 
-        if (ps.body == null && !hack) {
-            an = new StepAtomNode(exec, d, thread.head.get());
+        FlowNode an;
+        CpsThread thread = CpsThread.current();
+        boolean hasBody = false;
+        if (!hack) {
+            if (args != null && args.getClass().isArray()) {
+                int size = Array.getLength(args);
+                if (size > 0 && Array.get(args, size - 1) instanceof CpsClosure) {
+                    hasBody = true;
+                }
+            }
+        }
+        if (!hack && !hasBody) {
+            // Legacy Stage Step support means the step has no body but still takesImplicitBlockArgument
+            if (!(d instanceof StageStep.DescriptorImpl) && d.takesImplicitBlockArgument()) {
+                throw new IllegalStateException(String.format("%s step must be called with a body", name));
```

Comment:
Since this is user-facing and the stack trace is not helpful for understanding or solving the actual problem, I would throw `AbortException` instead so that the stack trace is not displayed if this exception causes the build to fail.

```suggestion
                throw new AbortException(String.format("%s step must be called with a body", name));
```

Reply:
So for an `AbortException`, I would have to update the signatures of `invokeStep`  and then catch it in `invokeMethod` because it's inherited  from `GroovyObjectSupport`. Should I just print an error  message to the log there and throw a `GroovyRuntimeException`?

Reply:
Ah, yeah, I forgot that you cannot throw `IOException` from these methods, so I would just leave it as-is. You could move things around and throw the exception by calling `CpsStepContext.onFailure` for the new step instead, but I think the way you have it is fine, and we can change it later if desired.